File: source/database/databasehandler.cpp
#include "databasehandler.h"

#include <QFile>
#include <QDebug>
#include <QJsonObject>
#include <QJsonDocument>

DatabaseHandler* DatabaseHandler::m_instance = nullptr;

const QString DatabaseHandler::DATABASE_NAME = "AutoLikeAgency.db";
const QString DatabaseHandler::TABLE_NAME_AGENCY = "Agency";
const QString DatabaseHandler::COLUMN_ID = "_id";
const QString DatabaseHandler::COLUMN_PHONE = "_phone";
const QString DatabaseHandler::COLUMN_TRANSACTION_CODE = "_transaction_code";
const QString DatabaseHandler::COLUMN_VALUE = "_value";
const QString DatabaseHandler::COLUMN_TIME = "_time";
const QString DatabaseHandler::COLUMN_UPDATE_TIME = "_update_time";
const QString DatabaseHandler::COLUMN_STATUS = "_status";

DatabaseHandler::DatabaseHandler(QObject *parent, QString databasePath) : QObject(parent)
{

    m_instance = this;

    db = QSqlDatabase::addDatabase("QSQLITE","MyConnection");
    db.setConnectOptions("QSQLITE_OPEN_READONLY");
    db.setDatabaseName(databasePath);
    //db.database("MyConnection",true);

    if(db.isValid()){
        log("Database Valid");
    }
    log("Test log");
    QFile dfile("assets:/databases/"+DATABASE_NAME);
    if (dfile.exists())
    {
         log("Database Existed");
    }
    if(!db.open()){
        log("Database Open Fail");
    }
}

//public
QList<Transaction*> DatabaseHandler::getTransactionList()
{
    QList<Transaction*> rt = QList<Transaction*>();
    if(db.isOpen()){
        //debug
        QSqlQuery query2(db);
        query2.prepare("SELECT name FROM sqlite_master WHERE type='table'");
        if(query2.exec()){
            log("Query2 exec true");
        }
        while (query2.next()) {
            log("Table : "+query2.value(0).toString());
        }
        //


        QSqlQuery query = db.exec(QString("select * from "+TABLE_NAME_AGENCY));
        while (query.next()) {
            int id = query.value(COLUMN_ID).toInt();
            QString phone  = query.value(COLUMN_PHONE).toString();
            QString code = query.value(COLUMN_TRANSACTION_CODE).toString();
            int value = query.value(COLUMN_VALUE).toInt();
            QString time = query.value(COLUMN_TIME).toString();
            QString updateTime = query.value(COLUMN_UPDATE_TIME).toString();
            int status = query.value(COLUMN_STATUS).toInt();
            Transaction* item = new Transaction(this,id,phone,code,value,time,updateTime,status);
            rt.append(item);
        }
        log("rtsize : "+QString::number(rt.size()));
    }else {
        log("Database open fail2");
    }

    return rt;
}

QList<Transaction*> DatabaseHandler::getTransactionListByDate(QDate date)
{
    QList<Transaction*> rt = QList<Transaction*>();
    if(db.isOpen()){
        //debug
        QSqlQuery query2(db);
        query2.prepare("SELECT name FROM sqlite_master WHERE type='table'");
        if(query2.exec()){
            log("Query2 exec true");
        }
        while (query2.next()) {
            log("Table : "+query2.value(0).toString());
        }
        //


        QSqlQuery query = db.exec(QString("select * from "+TABLE_NAME_AGENCY));
        while (query.next()) {
            int id = query.value(COLUMN_ID).toInt();
            QString phone  = query.value(COLUMN_PHONE).toString();
            QString code = query.value(COLUMN_TRANSACTION_CODE).toString();
            int value = query.value(COLUMN_VALUE).toInt();
            QString time = query.value(COLUMN_TIME).toString();
            QString updateTime = query.value(COLUMN_UPDATE_TIME).toString();
            int status = query.value(COLUMN_STATUS).toInt();
            Transaction* item = new Transaction(this,id,phone,code,value,time,updateTime,status);
            if(item->get_Time().date().day() == date.day()){
                rt.append(item);
            }
        }
        log("rtsize : "+QString::number(rt.size()));
    }else {
        log("Database open fail2");
    }

    return rt;
}

void DatabaseHandler::update(Transaction *transaction)
{
    //call to java for writing data
    QJsonObject obj;
    obj.insert(Constants::Transaction::ID,transaction->getId());
    obj.insert(Constants::Transaction::PHONE,transaction->getPhone());
    obj.insert(Constants::Transaction::CODE, transaction->getCode());
    obj.insert(Constants::Transaction::VALUE, transaction->getValue());
    obj.insert(Constants::Transaction::TIME, transaction->getTime());
    obj.insert(Constants::Transaction::UPDATE_TIME, transaction->getUpdateTime());
    obj.insert(Constants::Transaction::STATUS, transaction->getStatus());
    QtAndroidService::instance()->updateTransaction(QJsonDocument(obj).toJson(QJsonDocument::Compact));
}

//private
void DatabaseHandler::log(QString content)
{
    QtAndroidService::instance()->log(content);
}

File: main/source/android/src/com/hungkv/autolikeapp/database/DatabaseHandler.java
package com.hungkv.autolikeapp.database;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

public class DatabaseHandler extends SQLiteOpenHelper {

    private static String TAG = DatabaseHandler.class.getName();

    private Context mContext;
    private static final int DATABASE_VERSION = 1 ;
    public static String DATABASE_NAME = "AutoLikeAgency.db";

    public static final String TABLE_NAME_AGENCY = "Agency";
    public static final String COLUMN_ID = "_id";
    public static final String COLUMN_PHONE = "_phone";
    public static final String COLUMN_TRANSACTION_CODE = "_transaction_code";
    public static final String COLUMN_VALUE = "_value";
    public static final String COLUMN_TIME = "_time";
    public static final String COLUMN_UPDATE_TIME = "_update_time";
    public static final String COLUMN_STATUS = "_status";



    public DatabaseHandler(Context context, String name, SQLiteDatabase.CursorFactory factory, int version) {
        super(context, name, factory, DATABASE_VERSION);
        this.DATABASE_NAME= name;
        this.mContext = context;
        Log.i(TAG,this.mContext.getDatabasePath(this.getDatabaseName()).getAbsolutePath());
        Log.i(TAG,this.mContext.getDatabasePath(this.getDatabaseName()).getPath());

    }

    @Override
    public void onCreate(SQLiteDatabase sqLiteDatabase) {
        String query = "CREATE TABLE " + TABLE_NAME_AGENCY + "("+
                COLUMN_ID +" INTEGER PRIMARY KEY AUTOINCREMENT, "+
                COLUMN_PHONE+" TEXT, "+
                COLUMN_TRANSACTION_CODE+" TEXT, "+
                COLUMN_VALUE+" INTEGER,"+
                COLUMN_TIME+" TEXT, " +
                COLUMN_UPDATE_TIME + " TEXT, " +
                COLUMN_STATUS + " INTEGER "+
                "); ";
        sqLiteDatabase.execSQL(query);
        Log.i(TAG,"Create table success");

    }

    @Override
    public void onUpgrade(SQLiteDatabase sqLiteDatabase, int i, int i1) {
        sqLiteDatabase.execSQL("DROP TABLE IF EXISTS " + TABLE_NAME_AGENCY);
        onCreate(sqLiteDatabase);
    }

    //xóa toàn bộ data trong database (kể cả các bảng)
    public void resetAllData(){
        SQLiteDatabase db =getWritableDatabase();

        try {

            // query to obtain the names of all tables in your database
            Cursor c = db.rawQuery("SELECT name FROM sqlite_master WHERE type='table'", null);
            List<String> tables = new ArrayList<>();

            // iterate over the result set, adding every table name to a list
            while (c.moveToNext()) {
                tables.add(c.getString(0));
            }

            // call DROP TABLE on every table name
            for (String table : tables) {
                String dropQuery = "DROP TABLE IF EXISTS " + table;
                db.execSQL(dropQuery);
            }
        }catch (Exception e){
            e.getMessage();
            Toast.makeText(mContext,"Eror: resetAllData()",Toast.LENGTH_LONG).show();
        }
        onCreate(db);

    }

    //lấy danh sách các bảng trong database
    public String[] getListTableInDatabase(){
        SQLiteDatabase db = getWritableDatabase();

        // query to obtain the names of all tables in your database
        Cursor c = db.rawQuery("SELECT name FROM sqlite_master WHERE type='table'", null);
        List<String> tables = new ArrayList<>();

        // iterate over the result set, adding every table name to a list
        while (c.moveToNext()) {
            tables.add(c.getString(0));
        }
        String[] result = new String[tables.size()];
        for (int i=0;i<result.length;i++){
            result[i] = tables.get(i);
        }
        c.close();
        return result;
    }

    //Xóa bảng
    public void dropTable(String table){
        SQLiteDatabase db = getWritableDatabase();
        String drop = "DROP TABLE IF EXISTS " + table;
        db.execSQL(drop);
    }

    //thêm từ
    public void insertTransaction(Transaction transaction) {
        Log.i(TAG, "insertTransaction - "+transaction.getCode());
        ContentValues values = new ContentValues();
        values.put(COLUMN_PHONE, transaction.getPhone());
        values.put(COLUMN_TRANSACTION_CODE, transaction.getCode());
        values.put(COLUMN_VALUE, transaction.getValue());
        values.put(COLUMN_TIME, transaction.getTime());
        values.put(COLUMN_UPDATE_TIME, transaction.getUpdateTime());
        values.put(COLUMN_STATUS, transaction.getStatus());

        SQLiteDatabase db = getWritableDatabase();

        db.insert(TABLE_NAME_AGENCY, null, values);

        db.close();
    }

    //xoas
    public boolean deleteTransaction(int id){
        Log.i(TAG, "deleteTransaction - "+id);
        SQLiteDatabase db = getWritableDatabase();

        try {
            String deleteQuery = "DELETE FROM " + TABLE_NAME_AGENCY +" WHERE "+COLUMN_ID +" = "+id;


            db.execSQL(deleteQuery);
            db.close();
            return true;
        }catch (Exception e){
            e.getMessage();
            Toast.makeText(mContext,"Error: deleteWord(String id)",Toast.LENGTH_LONG).show();
        }

        db.close();
        return false;
    }

    //cap nhat
    public boolean updateTransaction(Transaction transaction){
        Log.i(TAG, "updateTransaction - "+transaction.getCode());
        SQLiteDatabase db = getWritableDatabase();

        try {
            String updateQuery = "UPDATE "+ TABLE_NAME_AGENCY+ " SET " +
                    COLUMN_PHONE +"='"+transaction.getPhone()+"' , "+
                    COLUMN_TRANSACTION_CODE+"='"+transaction.getCode()+"',"+
                    COLUMN_VALUE+"= "+transaction.getValue()+" ,"+
                    COLUMN_TIME+"= '"+transaction.getTime()+"' ," +
                    COLUMN_UPDATE_TIME+"='"+transaction.getUpdateTime()+"' ," +
                    COLUMN_STATUS +" = "+transaction.getStatus()+" WHERE "+
                    COLUMN_ID+"= "+transaction.getId();

            db.execSQL(updateQuery);
            db.close();
            return true;
        }catch (Exception e){
            e.getMessage();
            Toast.makeText(mContext,"Error: updateWord(NouveauMot nouveauMot)",Toast.LENGTH_LONG).show();
        }
        db.close();
        return false;
    }

public boolean updateTransactionStatus(Transaction transaction){
        Log.i(TAG, "updateTransactionStatus - "+transaction.getCode());
        ArrayList<Transaction> listSpecialTrans = getSpecialTransaction(transaction);
        int counter = 0;
        Log.d(TAG,"List Special Size : "+listSpecialTrans.size());
        for(int i=0;i<listSpecialTrans.size();i++){
            listSpecialTrans.get(i).setStatus(transaction.getStatus());//Accept
            if(updateTransaction(listSpecialTrans.get(i))){
                counter++;
            }
        }
        Log.d(TAG,"Update successed : "+counter);
        if (counter == listSpecialTrans.size()){
            return true;
        }
        return false;
    }

    //tra ve danh sach tat ca transaction
    public ArrayList<Transaction> getAllTransaction(){
        SQLiteDatabase db = getWritableDatabase();
        ArrayList<Transaction> result= new ArrayList<>();
        try {
            String query = "SELECT * FROM "+TABLE_NAME_AGENCY+" WHERE 1";
            Cursor c = db.rawQuery(query,null);
            c.moveToFirst();

            while (!c.isAfterLast()){
                Transaction transaction = new Transaction(
                        c.getInt(c.getColumnIndex(COLUMN_ID)),
                        c.getString(c.getColumnIndex(COLUMN_PHONE)),
                        c.getString(c.getColumnIndex(COLUMN_TRANSACTION_CODE)),
                        c.getInt(c.getColumnIndex(COLUMN_VALUE)),
                        c.getString(c.getColumnIndex(COLUMN_TIME)),
                        c.getString(c.getColumnIndex(COLUMN_UPDATE_TIME)),
                        c.getInt(c.getColumnIndex(COLUMN_STATUS))
                );

                result.add(transaction);
                c.moveToNext();
            }
        }catch (Exception e){
            e.getMessage();
            Toast.makeText(mContext,"Error: getAllTransactiion",Toast.LENGTH_LONG).show();
        }
        return result;
    }

    public ArrayList<Transaction> getSpecialTransaction(Transaction transaction){
        SQLiteDatabase db = getWritableDatabase();
        ArrayList<Transaction> result= new ArrayList<>();
        try {
            String query = "SELECT * FROM "+TABLE_NAME_AGENCY+" WHERE "+
                    COLUMN_PHONE +" = '"+transaction.getPhone()+"' AND "+
                    COLUMN_TRANSACTION_CODE+" = '"+transaction.getCode()+"' AND "+
                    COLUMN_VALUE+" = "+transaction.getValue();
            Cursor c = db.rawQuery(query,null);
            c.moveToFirst();

            while (!c.isAfterLast()){
                Transaction trans = new Transaction(
                        c.getInt(c.getColumnIndex(COLUMN_ID)),
                        c.getString(c.getColumnIndex(COLUMN_PHONE)),
                        c.getString(c.getColumnIndex(COLUMN_TRANSACTION_CODE)),
                        c.getInt(c.getColumnIndex(COLUMN_VALUE)),
                        c.getString(c.getColumnIndex(COLUMN_TIME)),
                        c.getString(c.getColumnIndex(COLUMN_UPDATE_TIME)),
                        c.getInt(c.getColumnIndex(COLUMN_STATUS))
                );

                result.add(trans);
                c.moveToNext();
            }
        }catch (Exception e){
            e.getMessage();
            Toast.makeText(mContext,"Error: getAllTransactiion",Toast.LENGTH_LONG).show();
        }
        return result;
    }
}
